Add out-of-bounds memory checking for GPUMem buffers and sub-buffers

JonathanLichtnerAMD opened the pull request:
This PR adds two different detection mechanisms for detecting out-of-bounds GPU buffer
memory accesses:
1. The first adds a new argument to MIOPenDriver, --gpubuffer_check <value>, which can be used to make the GPUMem class allocate GPU buffers that are aligned to 2 MiB alignment boundaries.  Specifying 1 for this option causes the client memory to be allocated at the start of the buffer (to detect OOBs before the buffer's start) and specifying 2 for this option causes the client memory to be aligned at the end of the buffer (to detect OOBs after the buffer's end).  Note that this is currently enabled for convolutions and batchnorm, although it is simple to add it for other problems.  (We might also want to consider turning this checking on by default since this only affects MIOpenDriver usage.)
2. The second method detects OOBs memory accesses for sub-buffers.  This can be enabled via the MIOPEN_DEBUG_CHECK_SUB_BUFFER_OOB_MEMORY_ACCESS environment variable, and uses a technique similar to that used for the GPUMem class (but adapted for sub-buffers).

JonathanLichtnerAMD committed:
Detect out-of-bounds memory accesses for MIOpen sub-buffers

This commit implements out-of-bounds memory detection for
MIOpen sub-buffers, which are buffers created from the workspace; that
is, the workspace is chopped into logical sub-buffers that can then be
used by the kernels.

Note that the previous detection methods might not detect invalid uses
of sub-buffers, since these sub-buffers were still part of the
workspace, and so going over the edge of a sub-buffer might be hidden
by the fact that the access was *into* a different sub-buffer.  This
is still an issue, but would not cause a GPU memory fault.

This commit introduces a new environment variable,
MIOPEN_DEBUG_CHECK_SUB_BUFFERS, which can be used to enable
out-of-bound memory access detection for sub-buffers.  This is done by
hipMalloc'ing each sub-buffer (in CreateSubBuffer()) and then aligning
the sub-buffer to either the left or the right of the 2 MiB aligned
pages.

If the environment variable is undefined or set to zero, then no
sub-buffer out-of-bounds detection is done. Otherwise, setting the
environment variable to 1 aligns the buffer to the left, which will
detect out-of-bounds before the start of the sub-buffer, and setting
the environment variable to 2 aligns the buffer to the right, which
detects out-of-bounds at the end of the sub-buffer.

Note that this is intended for internal debugging only.

JonathanLichtnerAMD committed:
Detect out-of-bounds memory accesses for MIOpen GPUMem conv buffers

This commit implements out-of-bounds (OOBs) memory detection for
MIOPen GPUMem convolution buffers.

This detection relies on the fact that the GPU will raise a memory
fault if an access into GPU buffers crosses a 2 MiB alignment
boundary *and* the memory adjacent to the boundary is not allocated, i.e.,
if the gpu memory is aligned to a 2 MiB page size boundary and the
memory access is either before the start of the buffer or after the
end of the buffer.

To make the problem more likely to occur, the memory returned to the
client is either allocated at the start of the buffer (to catch issues
before the start of the buffer) or at the end of the buffer (to catch
issues at the end of the buffer).

This detection is off by default, but can be enabled via
the MIOpenDriver --gpubuffer_check flag which takes the following
values:

    0 - no out-of-bound detection (default)
    1 - align the user-memory to the left (detect OOB before the start
        of the buffer)
    2 - align the user-memory to the right (detect OOB after the end
        of the buffer)

Note that this initial commit only enables this detection for
convolutions.

JonathanLichtnerAMD committed:
Add out-of-bounds checking for MIOpen batchnorm buffers

cderb commented on driver/driver.hpp:
```diff
@@ -131,8 +140,33 @@ struct GPUMem
+        constexpr size_t memoryPageSize = 2 * 1024 * 1024;
+
+        auto roundUp = [&](size_t bytes) {
+            return ((bytes + memoryPageSize) / memoryPageSize) * memoryPageSize;
```
This will count an extra page when bytes is exactly the page size.

BrianHarrisonAMD replied:
Originally in my example it allowed some extra room to help figure out how far it was going past the end, but in this version I don't think we need that. 

`((bytes + memoryPageSize - 1) / memoryPageSize) * memoryPageSize;`

could be used to avoid an extra page.

JonathanLichtnerAMD replied:
This is fixed.  Thanks for catching this!

cderb reviewed (approved):
LGTM

averinevg commented on driver/driver.hpp:
```diff
@@ -131,8 +140,33 @@ struct GPUMem
 
+    size_t GetTotalSize(size_t userSize)
+    {
+        constexpr size_t memoryPageSize = 2 * 1024 * 1024;
```
Why was this value chosen?
I think it is better to change the variable name to avoid confusion.

BrianHarrisonAMD replied:
Originally in my example it allowed some extra room to help figure out how far it was going past the end, but in this version I don't think we need that. 

`((bytes + memoryPageSize - 1) / memoryPageSize) * memoryPageSize;`

could be used to avoid an extra page.

JonathanLichtnerAMD replied:
By using 2 MiB, this causes the GPU to raise a memory fault if the access is across a 2 MiB boundary.  I've updated the name of the lambda and the variable to make this clearer.

averinevg commented on driver/driver.hpp:
```diff
@@ -452,6 +501,36 @@ inline void Driver::InitDataType()
+        exit(EXIT_FAILURE);
+    }
+}
+
```
This code should probably be placed in `driver.cpp`

JonathanLichtnerAMD replied:
Done

averinevg commented on src/hip/handlehip.cpp:
```diff
@@ -795,16 +796,123 @@ std::ostream& Handle::Print(std::ostream& os) const
+    void* mem            = nullptr;
+    auto status          = hipMalloc(&mem, totalSize);
+    if(status != hipSuccess)
+        MIOPEN_THROW_HIP_STATUS(status, "[MIOpenDriver] hipMalloc failed " + std::to_string(size));
```
`MIOpenDriver` is misleading here

JonathanLichtnerAMD replied:
Fixed.  Thanks for catching this!

averinevg replied:
I meant that this text shouldn't be there

JonathanLichtnerAMD replied:
Yes, I think I missed a few spots :)  I've removed all the MIOpenDriver text in this file now.

averinevg commented on src/hip/handlehip.cpp:
```diff
@@ -65,6 +65,7 @@
 #define WORKAROUND_FAULTY_HIPMEMGETINFO_VEGA_NAVI2X (HIP_PACKAGE_VERSION_FLAT >= 5007000000ULL)
 
 MIOPEN_DECLARE_ENV_VAR_UINT64(MIOPEN_DEVICE_CU)
+MIOPEN_DECLARE_ENV_VAR_UINT64(MIOPEN_DEBUG_CHECK_SUB_BUFFERS)
```
From the name it is not entirely clear what check is being performed. It would be good to clarify this it the name and also add it to the documentation. Additionally it should be noted that no actual check is being performed.

JonathanLichtnerAMD replied:
Yes, the check is implicit.  How about MIOPEN_DEBUG_CHECK_SUB_BUFFER_OOB_MEMORY_ACCESS?  It's a bit long but it's more descriptive....

JonathanLichtnerAMD replied:
I've renamed it and added to the documentation

averinevg commented on driver/driver.hpp:
```diff
@@ -131,8 +140,33 @@ struct GPUMem
+        constexpr size_t memoryPageSize = 2 * 1024 * 1024;
+
+        auto roundUp = [&](size_t bytes) {
+            return ((bytes + memoryPageSize) / memoryPageSize) * memoryPageSize;
```
When `bytes` is close to `memoryPageSize`, all of this becomes meaningless.

BrianHarrisonAMD replied:
Yea if it's right on the page size, then it's basically a no-op. 
This just gives us a way to verify memory access issues for kernels that can't use ASAN.

averinevg replied:
When it equals the page size, we have a one page buffer, but if it's smaller, this will be a problem, as it will simply not enough to compensate memory access issues. I also have concerns about unaligned data access. Wouldn't be better to simply add a page before or after the buffer? This way, we will always have a buffer with a constant size of reserved space.

BrianHarrisonAMD replied:
I think we want the buffer to be aligned with the end or start of a page boundary to force consistently triggering memory faults. Since ASAN doesn't work with ASM kernels, I think we want this as a way to help reproduce memory access issues when they come up.

The alignment issues are interesting. I think we should be also checking for that, and failing gracefully if it's not a valid alignment for the kernel. Unless the expectation is that all the memory is aligned properly, then we should modify this to match that.

How would adding a page before & after help with detection?

JonathanLichtnerAMD replied:
The alignment issue is interesting.  The sub-buffers were padded out to a 256 byte alignment, which this change loses when we put the client buffer at the end of the larger buffer.  We could potentially add this padding back in but this would also possibly hide some OOB access issues (since reads or writes into the padding wouldn't be detected).  We could also add a similar padding for the GPUMem checking if we feel it's warranted.  However, Chris ran Brian's changes on a great number of shapes and I don't think he saw any failures that were due to the alignment.  But maybe we just didn't notice these because the memory fault issues were much more glaring :) 

@averinevg do we know the impact of this alignment.  I suspect this might affect performance, but I am new to GPU memory and kernel debugging.

averinevg replied:
I finally get how this works.

As for alignment, I think aligning to a large power of 2 should be sufficient; however, I believe this might potentially affect performance.

averinevg replied:
Resolved

BrianHarrisonAMD commented on driver/driver.hpp:
```diff
@@ -131,8 +140,33 @@ struct GPUMem
+            return ((bytes + memoryPageSize) / memoryPageSize) * memoryPageSize;
+        };
+
+        if(check == Check::None)
+            return userSize;
```
Nit: I think having this at the top of the function makes it a bit more clear.

JonathanLichtnerAMD replied:
Fixed.   Thanks!

JonathanLichtnerAMD committed:
Move the handling of the Check::None condition to the top of GetTotalSize()

JonathanLichtnerAMD committed:
Fix overallocation issue when the size is a multiple of 2 MiB

Also rename some variables and lambdas to help clarify the
intention of the code.

JonathanLichtnerAMD committed:
Fix some warnings from the Hip Tidy CI job

JonathanLichtnerAMD committed:
Move the two new Driver methods to driver.cpp

JonathanLichtnerAMD committed:
Rename MIOPEN_DEBUG_CHECK_SUB_BUFFER to MIOPEN_DEBUG_CHECK_SUB_BUFFER_OOB_MEMORY_ACCESS

JonathanLichtnerAMD committed:
Add documentation for MIOPEN_DEBUG_CHECK_SUB_BUFFER_OOB_MEMORY_ACCESS

JonathanLichtnerAMD committed:
Fix formatting

amd-jnovotny commented on docs/how-to/debug-log.rst:
```diff
@@ -319,6 +319,18 @@ settings to help with debugging:
+==========================================================
+Use the ``MIOPEN_DEBUG_CHECK_SUB_BUFFER_OOB_MEMORY_ACCESS`` to help debug potential out-of-bounds (OOBs)
+memory access errors on GPU sub-buffers.  If the environment variable is undefined or set to zero, then no
+sub-buffer out-of-bounds detection is done.  Otherwise:
```
```suggestion
sub-buffer out-of-bounds detection is performed. To verify the memory accesses, use one of these values:
```

amd-jnovotny commented on docs/how-to/debug-log.rst:
```diff
@@ -319,6 +319,18 @@ settings to help with debugging:
+* ``MIOPEN_DEBUG_CHECK_SUB_BUFFER_OOB_MEMORY_ACCESS=1``: Check for OOBs before the start of the sub-buffers.
+* ``MIOPEN_DEBUG_CHECK_SUB_BUFFER_OOB_MEMORY_ACCESS=2``: Check for OOBs after the end of the sub-buffers.
+
+Note that this checking is not intended for production use since there is a possible performance hit
+from enabling this functionality.
```
```suggestion
Note that these checks are not intended for production use because they might cause a performance hit.
```

JonathanLichtnerAMD replied:
I've made these changes.  Thanks!

amd-jnovotny reviewed (approved):
Just a couple of suggestions for the docs.

JonathanLichtnerAMD committed:
Cleanup usage of MIOpenDriver logging in handlehip.cpp

JonathanLichtnerAMD committed:
Update docs for MIOPEN_DEBUG_CHECK_SUB_BUFFER_OOB_MEMORY_ACCESS

averinevg reviewed (approved):
LGTM

randyspauldingamd commented on driver/bn_driver.hpp:
```diff
@@ -526,31 +531,36 @@ int BatchNormDriver<TInput, Tref, TAcc, TScaleBias, TOut>::AllocateBuffersAndCop
     clGetCommandQueueInfo(q, CL_QUEUE_CONTEXT, sizeof(cl_context), &ctx, nullptr);
 #endif
-    status |= in.AllocOnDeviceAndInit(q, ctx, GetTensorSize(&in.GetTensor().desc));
+    status |= in.AllocOnDeviceAndInit(q, ctx, GetTensorSize(&in.GetTensor().desc), buffer_check);
```
I think it's important to add unit tests for this so we know how successful it is. I started working out how to do that to give you a head start, and haven't quite got it working yet. Let's discuss before it goes in.

randyspauldingamd replied:
Jonathan will add the Unit Tests in future PRs.

JonathanLichtnerAMD replied:
Yes, unit tests for this would be great.  I've created story https://ontrack-internal.amd.com/browse/LWPMIOPEN-1443 to track the work to add unit tests.

JonathanLichtnerAMD committed:
Merge branch 'develop' into jlichtne/memfault-memcheck
